### PR TITLE
Detect Expressions that evaluate to Unit and warn the user

### LINKED ISSUE #2011: Version of let_unit_value that is an early pass
For https://github.com/rust-lang/rust/issues/44173

We should have a "use unit value" lint that checks for:

 - `()` being assigned to a variable
 - `()` being passed to a function

where the origin of the `()` is basically a semicolonless block

An example of an early pass is https://github.com/rust-lang-nursery/rust-clippy/blob/master/clippy_lints/src/precedence.rs

What we want is an early pass (runs on the AST before typechecking) with a `check_expr` and `check_stmt` that look for:

 - [assignment expressions](http://manishearth.github.io/rust-internals-docs/syntax/ast/enum.ExprKind.html#Assign.v)
 - [let statements](http://manishearth.github.io/rust-internals-docs/syntax/ast/enum.StmtKind.html#Local.v)
 - [MethodCall expressions](http://manishearth.github.io/rust-internals-docs/syntax/ast/enum.ExprKind.html#MethodCall.v)
 - [Call expressions](http://manishearth.github.io/rust-internals-docs/syntax/ast/enum.ExprKind.html#Call.v)

and then run a `is_unit_expr` check on their contents (the initializer for assignment/let, and each of the arguments for calls)

`is_unit_expr` will:

 - if it's a [`Block`](http://manishearth.github.io/rust-internals-docs/syntax/ast/enum.ExprKind.html#Block.v), return true if the last statement is not an [`Expr` statement](http://manishearth.github.io/rust-internals-docs/syntax/ast/enum.StmtKind.html#Expr.v)
 - if it's an [if expr](http://manishearth.github.io/rust-internals-docs/syntax/ast/enum.ExprKind.html#If.v) that has an else block, return true if both the if and else blocks have a non-expr final statement
 - We can do something similar for match exprs if we wish.
 - return false otherwise

We can also have it recurse, where if the last statement _is_ a `StmtKind::Expr`, then recurse down and check if _that_ expr evaluates to unit.

Then, if any of the `is_unit_expr` calls evaluates to true, throw the lint.

@zmanian is working on this

### PR DESCRIPTION
**Pull request by zmanian**
Work in progress to resolve #2011 

Feedback requested.

### CODE REVIEW

File: clippy_lints/src/is_unit_expr.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            if local.pat.node == PatKind::Wild {return;}
+            if let Some(ref expr) = local.init{
+                if let Some(span) = is_unit_expr(expr){
+                    cx.span_lint(UNIT_EXPR, span, "Consider removing the trailing semicolon");  
+            }
+    }
+}
+}
+}
+fn is_unit_expr(expr: &Expr)->Option<Span>{
```
Comment (Manishearth) [inline clippy_lints/src/is_unit_expr.rs]: probably rustfmt the file

File: clippy_lints/src/is_unit_expr.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+         ExprKind::Block(ref block) => {
+             if check_last_stmt_in_block(block){
+                 return Some(block.stmts[block.stmts.len()-1].span.clone());
+             } else{
+                 return None;
+             }
+        },
+        ExprKind::If(_, ref then, ref else_)=>{
+            let check_then = check_last_stmt_in_block(then);
+            if let Some(ref else_) = *else_{
```
Comment (Manishearth) [inline clippy_lints/src/is_unit_expr.rs]: If there is no else block then it's always going to evaluate to `()`

File: clippy_lints/src/is_unit_expr.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                let check_else = is_unit_expr(else_.deref());
```
Comment (Manishearth) [inline clippy_lints/src/is_unit_expr.rs]: you shouldn't need to call deref directly. `&else_` should work here.

File: clippy_lints/src/is_unit_expr.rs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                    return Some(expr_else.clone());
+                }else{
+                    return Some(expr.span.clone());
+                }
+            } 
+            if check_then { 
+                return Some(expr.span.clone());
+
+            } else{
+                return Some(expr.span.clone());
```
Comment (Manishearth) [inline clippy_lints/src/is_unit_expr.rs]: This else block shouldn't be there I think? `return None`?

File: clippy_lints/src/is_unit_expr.rs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+             }
+        },
+        ExprKind::If(_, ref then, ref else_)=>{
+            let check_then = check_last_stmt_in_block(then);
+            if let Some(ref else_) = *else_{
+                let check_else = is_unit_expr(else_.deref());
+                if let Some(ref expr_else) = check_else{
+                    return Some(expr_else.clone());
+                }else{
+                    return Some(expr.span.clone());
```
Comment (Manishearth) [inline clippy_lints/src/is_unit_expr.rs]: `return None`?

File: clippy_lints/src/is_unit_expr.rs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+fn check_last_stmt_in_block(block: &Block)->bool{
+    let ref final_stmt = &block.stmts[block.stmts.len()-1];
+        if let StmtKind::Expr(_) = final_stmt.node{
+                return false;
+            }
+            else{
+              return true; 
+            }
+}
```
Comment (Manishearth) [inline clippy_lints/src/is_unit_expr.rs]: nit: newline at EOF

File: tests/ui/is_unit_expr.rs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+
+#![feature(plugin)]
+#[plugin(clippy)]
```
Comment (Manishearth) [inline tests/ui/is_unit_expr.rs]: Needs an exclamation mark. The tests didn't run correctly (see the output)

File: clippy_lints/src/is_unit_expr.rs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+            }
+        }
+        if let ExprKind::MethodCall(ref _left, ref args) = expr.node {
+            for ref arg in args {
+                if let Some(span) = is_unit_expr(arg) {
+                    span_note_and_lint(
+                        cx,
+                        UNIT_EXPR,
+                        expr.span,
+                        "This expression assigns the Unit type ()",
```
Comment (Manishearth) [inline clippy_lints/src/is_unit_expr.rs]: "assigns" probably isn't the right word to use here for a call

File: clippy_lints/src/is_unit_expr.rs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+            }
+        }
+        if let ExprKind::Call(_, ref args) = expr.node {
+            for ref arg in args {
+                if let Some(span) = is_unit_expr(arg) {
+                    span_note_and_lint(
+                        cx,
+                        UNIT_EXPR,
+                        expr.span,
+                        "This expression assigns the Unit type ()",
```
Comment (Manishearth) [inline clippy_lints/src/is_unit_expr.rs]: ditto

**Comment by Manishearth** [comment]
Needs rustfmt on the tests, otherwise it's good to go.

It occurs to me that we have chances for false positives on if blocks that diverge via panics too.

Perhaps we should only lint if _both_ the if and else are semicolon'd?

**Comment by Manishearth** [comment]
or, well, at worst it will warn about a needless semicolon after a diverging statement. That's fine.

**Comment by Manishearth** [comment]
retriggering travis

**Comment by Arnavion** [comment]
>or, well, at worst it will warn about a needless semicolon after a diverging statement. That's fine.

Does "That's fine." mean it's fine *for now* and that it's planned to be fixed? Or it's by design? I like putting a semi-colon after my `continue`s and `break`s and `return`s :\

**Comment by oli-obk** [comment]
We can detect `continue`, `break` and `return` without troubles. It's `foo()` where `fn foo() -> !` that's problematic.

**Comment by oli-obk** [comment]
`retutrn` and `break` actually don't trigger this, just `continue`, opening an issue

**Comment by Arnavion** [comment]
Ah cool. I did only hit the false positive with `continue`

**Comment by zmanian** [comment]
Oh continue is an easy fix.

**Comment by oli-obk** [comment]
Fixed in https://github.com/rust-lang-nursery/rust-clippy/pull/2022

**Comment by Manishearth** [comment]
> Does "That's fine." mean it's fine for now and that it's planned to be fixed?

I meant that it's fine forever.

`continue` was an omission, that comment was about diverging functions.

**Pull request merged**